modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersValidator.java

```diff
@@ -0,0 +1,168 @@
+     * {@link #getValidatorInboundHandler()} can use to convey the validation result.
+     */
+    public static HttpMessage wrapAsValidatableMessage(HttpMessage newlyDecodedMessage) {
+        assert newlyDecodedMessage instanceof HttpRequest;
+        DefaultHttpRequest httpRequest = (DefaultHttpRequest) newlyDecodedMessage;
+        ValidatableHttpHeaders validatableHttpHeaders = new ValidatableHttpHeaders(newlyDecodedMessage.headers());
+        return new DefaultHttpRequest(httpRequest.protocolVersion(), httpRequest.method(), httpRequest.uri(), validatableHttpHeaders);
+    }
+
+    public static ValidationResult extractValidationResult(org.elasticsearch.http.HttpRequest request) {
+        ValidatableHttpHeaders authenticatedHeaders = unwrapValidatableHeaders(request);
+        return authenticatedHeaders != null ? authenticatedHeaders.validationResultContextSetOnce.get() : null;
+    }
+
+    private static ValidatableHttpHeaders unwrapValidatableHeaders(org.elasticsearch.http.HttpRequest request) {
+        if (request instanceof Netty4HttpRequest == false) {
+            return null;
+        }
+        if (((Netty4HttpRequest) request).getNettyRequest().headers() instanceof ValidatableHttpHeaders == false) {
+            return null;
+        }
+        return (ValidatableHttpHeaders) (((Netty4HttpRequest) request).getNettyRequest().headers());
+    }
+
+    /**
+     * {@link HttpHeaders} implementation that carries along the {@link ValidationResult} iff
+     * the HTTP headers have been validated successfully.
+     */
+    public static final class ValidatableHttpHeaders extends DefaultHttpHeaders {
+
+        @FunctionalInterface
+        public interface ValidationResult {
+            void assertValid();
+        }
+
+        public final SetOnce<ValidationResult> validationResultContextSetOnce;
+
+        public ValidatableHttpHeaders(HttpHeaders httpHeaders) {
+            this(httpHeaders, new SetOnce<>());
+        }
+
+        private ValidatableHttpHeaders(HttpHeaders httpHeaders, SetOnce<ValidationResult> validationResultContextSetOnce) {
+            // the constructor implements the same logic as HttpHeaders#copy
+            super();
+            set(httpHeaders);
+            this.validationResultContextSetOnce = validationResultContextSetOnce;
+        }
+
+        private ValidatableHttpHeaders(HttpHeaders httpHeaders, ValidationResult validationResult) {
+            this(httpHeaders);
+            if (validationResult != null) {
+                markAsSuccessfullyValidated(validationResult);
+            }
+        }
+
+        /**
+         * Must be called at most once in order to mark the http headers as successfully validated.
+         * The intent of the {@link ValidationResult} parameter is to associate some details about the validation
+         * that can be later retrieved when dispatching the request.
+         */
+        public void markAsSuccessfullyValidated(ValidationResult validationResult) {
```

Comment:
can this be `addAuthContext(ThreadContext.StoredContext authContext)` (not specifically those names, just trying avoid skirting around that this is the result of the authentication)

Reply:
I pushed be7066451beabbfbf10c1789a20a30989105e4b2 to rename to `addValidationContext`.

Reply:
After [625e56c](https://github.com/elastic/elasticsearch/pull/95112/commits/625e56c38b302bd5f7e788ad18f0e183e5c40f1a) it has become `setAuthenticationContext`.